### DISCUSSION TITLE
RTL Support

### DISCUSSION THREAD

**Discussion by 3adeling** [tailwindlabs/tailwindcss#1492] [General]
Hey there

600 millions of people in the world write and read from RTL.

It is disappointing that this must-have feature is not supported yet in Tailwind (it doesn't even respect the html `dir="rtl"` attribute).

do you have any plan to support RTL anytime soon?

**Comment by adamwathan** [comment]
See previous discussion here 👍🏻

https://github.com/tailwindcss/discuss/issues/64#issuecomment-532801714

**Comment by 3adeling** [comment]
hey @adamwathan,

the plugin you mentioned doesn't work for me. I am using [Laravel Tailwind Preset](https://github.com/laravel-frontend-presets/tailwindcss). I am not sure whats wrong.

We, the community, can provide support in fixing RTL bugs and issues but at least we need an official starting point in that direction from Tailwind team.

Regards,

**Comment by adamwathan** [comment]
What doesn't work about it? Can you share an example project where you have tried to set it up so I can see what the issue is?

**Comment by 3adeling** [comment]
both `rtl:` & `ltr:` doesn't work.

I created new laravel project + [Laravel Tailwind Preset](https://github.com/laravel-frontend-presets/tailwindcss) + [Tailwind Dir](https://github.com/RonMelkhior/tailwindcss-dir):
https://github.com/3adeling/tailwind-rtl-demo

I only used `ltr:mr-5` & `rtl:ml-5` and both of them isn't working when changing `<html dir="ltr">`

**Comment by adamwathan** [comment]
Can you make the repo public so I can see?

**Comment by 3adeling** [comment]
It's public now. sorry for that :)

**Comment by adamwathan** [comment]
Seems to be working fine:

![image](https://user-images.githubusercontent.com/4323180/78055810-14b2dc00-7352-11ea-88e5-ddd33e20b5f7.png)
![image](https://user-images.githubusercontent.com/4323180/78055836-1d0b1700-7352-11ea-8ad8-0f7ea39c538c.png)

I noticed the CSS file in your `public` folder was old and hadn't been recompiled, so make sure you run `npm run dev` (or have `npm run watch` running) to rebuild your CSS when you make changes to your Tailwind config.

**Comment by 3adeling** [comment]
Thank you very much @adamwathan. it is working now.

do I have to run `npm run dev` whenerver I add a new plugin?

**Comment by adamwathan** [comment]
You have to rebuild your CSS files any time you make any sort of change that would result in different CSS yeah.

**Comment by 20lives** [comment]
In 99% of the cases the `rtl:` variant is trivial and can be predicted.

So it's really unreasonable to transform this:
```
<div class="text-green border pl-4 mr-1 right-0 rounded-l-md">
```
to this:
```
<div class="text-green border ltr:pl-4 ltr:mr-1 ltr:right-0 ltr:rounded-l-md rtl:pr-4 ltr:ml-1 rtl:left-0 rtl:rounded-r-md">
```
moreover it's much harder to read and maintain.

I have written another plugin with a different approach: [tailwindcss-rtl](https://github.com/20lives/tailwindcss-rtl)

just use **start** and **end** notations instead of **left** and **right** depends on your language direction.

Here is [Workcation](https://github.com/20lives/workcation-nuxt-i18n-rtl) transformed with a simple search and replace.

**Reply by jquense** [reply]
+1 to variants being a bad general solution to this problem. Since tailwind v3 is now JIT only, adding support for CSS logical properties would replace the need for tailwindcss-rtl, which is great, but doesn't seem easy to maintain!

**Comment by Reex11** [comment]
+1 to variants being a bad general solution to this problem.

I am currently working on Tailwind v2 and I'm using a plugin which added classes as the following:

`ms-x` Margin to Start ( Adds margin to the right if RTL and to the left otherwise )
`me-x` Margin to End ( Adds margin to the left if RTL and to the right otherwise )

I very much like that because it will automaticly decide the direction without having to specify both RTL and LTR. Also, I'll be able to force left or right if I want by using the default `ml-x` and `mr-x`.

Original Code:
```
<div class="text-green border pl-4 mr-1 right-0 rounded-l-md">
```

New Code:
```
<div class="text-green border ps-4 me-1 end-0 rounded-s-md">
```
This way we have RTL and the original code will work as expected but without RTL functionality.
Also, I can have only the padding to be left whatever is the layout direction:
```
<div class="text-green border pl-4 me-1 end-0 rounded-s-md">
```

**Reply by johanekhager** [reply]
+1. This solution make much more sense compare to adding `ltr` and `rtl` variants. Would love this solution. 🥳

**Reply by thibaudcolas** [reply]
See #1483, which discusses this approach in more details.

**Comment by korenLevi** [comment]
We will always need to add "ltr:pl-4" to our class ?

**Comment by alikazemkhanloo** [comment]
`postcss-rtlcss` plugin adds support for rtl/ltr. no need for extra `rtl:` prefix in class names. just change the `dir` attribute in `html` tag.

**Reply by aneesdev** [reply]
> `postcss-rtlcss` plugin adds support for rtl/ltr. no need for extra `rtl:` prefix in class names. just change the `dir` attribute in `html` tag.

Any example how to integrate that?

**Reply by alikazemkhanloo** [reply]
As simple as this:
```shell
npm i postcss-rtlcss
```
```js
// postcss.config.js
module.exports = {
    plugins: {
        tailwindcss: {},
        autoprefixer: {},    
        "postcss-rtlcss": {
         // invert ltr and rtl prefixes so that left and right would be natural in rtl
            ltrPrefix: '[dir="rtl"]',
            rtlPrefix: '[dir="ltr"]',
        }
    }
};
```
In my case, I was developing in RTL mode so I had to invert left and right to make it feel natural.

Another solution would be to use inline CSS properties as provided by the plugin `tailwindcss-rtl`. although haven't used it yet, but I prefer this one. Generates smaller CSS.

**Reply by aneesdev** [reply]
=

> As simple as this:
> 
> ```shell
> npm i postcss-rtlcss
> ```
> 
> ```js
> // postcss.config.js
> module.exports = {
>     plugins: {
>         tailwindcss: {},
>         autoprefixer: {},    
>         "postcss-rtlcss": {
>          // invert ltr and rtl prefixes so that left and right would be natural in rtl
>             ltrPrefix: '[dir="rtl"]',
>             rtlPrefix: '[dir="ltr"]',
>         }
>     }
> };
> ```
> 
> In my case, I was developing in RTL mode so I had to invert left and right to make it feel natural.
> 
> Another solution would be to use inline CSS properties as provided by the plugin `tailwindcss-rtl`. although haven't used it yet, but I prefer this one. Generates smaller CSS.

Thanks, man. really appreciate it

**Reply by HasanMothaffar** [reply]
Thank you! Saved me a lot of time searching for plugins.

**Reply by tatthang96** [reply]
@alikazemkhanloo Hi, thank you for your suggestion. However, I believe there is an issue when using it with translate because postcss-rtlcss doesn't seem to detect it.

**Comment by adamwathan** [comment]
Native logical properties support is coming tomorrow 🎁

**Comment by Aswinjithkukku** [comment]
what if we need to use screen size tags with it like md:pr-4 and how to add rtl to this

**Reply by HasanMothaffar** [reply]
You can use Tailwindcss's logical properties. Instead of using `md:pr-4`, use `md:pe-4` which works automatically in both LTR and RTL layouts. You can read more about logical properties here: https://tailwindcss.com/blog/tailwindcss-v3-3#simplified-rtl-support-with-logical-properties

**Reply by zoltanszogyenyi** [reply]
It's what we used over at Flowbite for our open-source UI components [RTL integration](https://flowbite.com/docs/customize/rtl/) as well and it has been a very seamless integration - so for sure logical properties are the way to go for Tailwind CSS.

**Comment by code-creeper** [comment]
I having the same issue then I found this in the docs
https://tailwindcss.com/docs/hover-focus-and-other-states#rtl-support

**Comment by mymarmalade** [comment]
I need simple tutorial about RTL website @adamwathan 
every where full of tutorial but empty of RTL Adam

**Comment by spacemudd** [comment]
When starting a new Laravel project, I used this prompt to update the quirks of RTL. Helps me a lot.

[laravel-rtl.md](https://github.com/user-attachments/files/23840899/laravel-rtl.md)
